Assertion failed when using QueryCursor::matches

### Problem

I generated a new parser using `tree-sitter generate`. I didn't make any other changes. Then I created a Rust project and tried to wire things up to call the `QueryCursor::matches` (or `QueryCursor::captures`) method and it failed.

The parsing part seems to be working fine. I also tried in Tree-sitter playground and below query works.

### Steps to reproduce

- Create a new parser with the default grammar:
_tree-sitter-example/grammar.js_
```javascript
/// <reference types="tree-sitter-cli/dsl" />
// @ts-check

module.exports = grammar({
  name: "example",

  rules: {
    // TODO: add the actual grammar rules
    source_file: $ => "hello"
  }
});
```
- Create a Rust project with the following dependencies:
_Cargo.toml_
```toml
[dependencies]
tree-sitter = "0.22.2"
tree-sitter-example = { path = "tree-sitter-example" }
```
- Run the following code:
_src/main.rs_
```rust
use tree_sitter::{Parser, Query, QueryCursor};

fn main() {
    let mut parser = Parser::new();
    let language = tree_sitter_example::language();
    parser.set_language(&language).unwrap();
    let query = Query::new(&language, r#""hello" @hello"#).unwrap();
    let text = "hello";
    let tree = parser.parse(text, None).unwrap();
    let mut query_matches = QueryCursor::new().matches(&query, tree.root_node(), text.as_bytes());
    query_matches.next();
}
```
- See the error:
```
~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tree-sitter-0.22.2/src/./tree_cursor.c:476: TSNode ts_tree_cursor_current_node(const TSTreeCursor *): Assertion `(uint32_t)((&self->stack)->size - 1) < (&self->stack)->size' failed.
```

### Expected behavior

The above example should work. Did I miss something?

### Tree-sitter version (tree-sitter --version)

tree-sitter 0.22.2

### Operating system/version

NixOS 24.05

This is really interesting! I took a look at this with a debugger and I think I found the problem. Namely, this line:
```rust
let mut query_matches = QueryCursor::new().matches(&query, tree.root_node(), text.as_bytes());
```

With the code written this way, `drop()` is called for the `QueryCursor` object. The `drop` implementation calls `ts_query_cursor_delete()` via ffi, which clears the contents of the array. Then, when you call `query_matches.next();`, the underlying cursor object has been deallocated, leading to the assertion triggering due to an array with zero length:

```c
/// Get a pointer to the element at a given `index` in the array.
#define array_get(self, _index) \
  (assert((uint32_t)(_index) < (self)->size), &(self)->contents[_index])
```
Splitting the cursor's creation and the `.matches()` call into separate lines, or just referencing the cursor later on to keep the object in scope for the `matches()` call can work around the problem. For a more permanent fix, adding a lifetime annotation for the cursor in `matches()`'s definition appears to fix things:

```rust
#[doc(alias = "ts_query_cursor_exec")]
pub fn matches<'query, 'cursor: 'query, 'tree, T: TextProvider<I>, I: AsRef<[u8]>>(
    &'cursor mut self,
    query: &'query Query,
    node: Node<'tree>,
    text_provider: T,
) -> QueryMatches<'query, 'tree, T, I> {
    let ptr = self.ptr.as_ptr();
    unsafe { ffi::ts_query_cursor_exec(ptr, query.ptr.as_ptr(), node.0) };
    QueryMatches {
        ptr,
        query,
        text_provider,
        buffer1: Vec::default(),
        buffer2: Vec::default(),
        _phantom: PhantomData,
    }
}
```
Attempting to build your file with the above changes to tree-sitter yields the following error, which seems to correctly describe the problem:
```bash
error[E0716]: temporary value dropped while borrowed
  --> src/main.rs:10:29
   |
10 |     let mut query_matches = QueryCursor::new().matches(&query, tree.root_node(), text.as_bytes());
   |                             ^^^^^^^^^^^^^^^^^^                                                   - temporary value is freed at the end of this statement
   |                             |
   |                             creates a temporary value which is freed while still in use
...
13 |     query_matches.next();
   |     ------------- borrow later used here
   |
help: consider using a `let` binding to create a longer lived value
   |
10 ~     let binding = QueryCursor::new();
11 ~     let mut query_matches = binding.matches(&query, tree.root_node(), text.as_bytes());
   |

For more information about this error, try `rustc --explain E0716`.
error: could not compile `ts_test` (bin "ts_test") due to 1 previous error
```

I'll open up a PR with the fix :)

--- a/lib/binding_rust/lib.rs
+++ b/lib/binding_rust/lib.rs
@@ -2336,8 +2336,8 @@ impl QueryCursor {
     /// Because multiple patterns can match the same set of nodes, one match may contain
     /// captures that appear *before* some of the captures from a previous match.
     #[doc(alias = "ts_query_cursor_exec")]
-    pub fn matches<'query, 'tree, T: TextProvider<I>, I: AsRef<[u8]>>(
-        &mut self,
+    pub fn matches<'query, 'cursor: 'query, 'tree, T: TextProvider<I>, I: AsRef<[u8]>>(
+        &'cursor mut self,
         query: &'query Query,
         node: Node<'tree>,
         text_provider: T,

